functorch/functorch/experimental/cond.py

```diff
@@ -0,0 +1,137 @@
+            proxy_out = e.__dict__[proxy_mode.tracer]
+            return proxy_out.proxy
+        return e
+
+
+    if isinstance(operands, torch.Tensor):
+        operands = [operands]  # Little hack because * on a single ProxyTensor unpacks it
+    else:
+        operands = operands
+
+    true_graph = get_isolated_graphmodule(true_fn, operands, {})
+    false_graph = get_isolated_graphmodule(false_fn, operands, {})
+    true_name = "true_graph"
+    false_name = "false_graph"
+    proxy_mode.tracer.root.register_module(true_name, true_graph)
+    proxy_mode.tracer.root.register_module(false_name, false_graph)
+
+    true_args = []
+    for node in true_graph.graph.nodes:
+        if node.op == 'output':
+            true_args.extend(*node.args)
+
+    false_args = []
+    for node in false_graph.graph.nodes:
+        if node.op == 'output':
+            false_args.extend(*node.args)
+
+    assert(len(true_args) == len(false_args))
+    for i in range(len(true_args)):
+        t_arg_meta = true_args[i]
+        f_arg_meta = false_args[i]
+        # WIP don't look at this yet
+        # print(t_arg_meta.meta)
+        # print(f_arg_meta.meta)
+        # assert(t_arg_meta.meta == f_arg_meta.meta)
+
+    args = (pred, true_graph, false_graph, operands)
+
+    proxy_args = pytree.tree_map(_unwrap_proxy, args)
+
+    # Does this need random slug appended so as not to collide?
+    proxy_res = proxy_mode.tracer.create_proxy('call_function', func_overload, proxy_args, kwargs,
+                                            name="conditional")
+    return proxy_res
+
+
+def cond_dense(pred, true_fn, false_fn, *operands):
+    mode = torch._C._get_torch_dispatch_mode()
+    assert (mode is None), "Mode should never be enabled for CPU key"
+    try:
+        if pred:
+            return true_fn(*operands)
+        else:
+            return false_fn(*operands)
+    except Exception as e:
+        # Do something proper here, someday
+        print("Exception:", e)
+        raise e
+
+
+def cond_autograd(pred, true_fn, false_fn, *operands):
+    # TODO: support autograd
+    flat_operands, _ = tree_flatten((true_fn, false_fn) + operands)
+    assert all([not f.requires_grad for f in flat_operands
+                if isinstance(f, torch.Tensor)])
+
+    guard = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.AutogradCPU))
+    return cond(pred, true_fn, false_fn, *operands)
+
+
+def cond_adinplaceorview(*args, **kwargs):
+    guard = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.ADInplaceOrView))
+    return cond(*args, **kwargs)
+
+def cond_backendselect(*args, **kwargs):
+    guard = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.BackendSelect))
+    return cond(*args, **kwargs)
+
+def python_fallback(op):
+    def inner(*args, **kwargs):
+        mode = torch._C._get_torch_dispatch_mode()
+        if mode:
+            with suspend_mode(mode):
+                res = trace_cond(mode, op, args, kwargs)
```

Comment:
Is something like the following reasonable:
```
mode.__torch_dispatch__(op, args, kwargs)
```
and always let users of `make_fx` decide the tracing behavior, like this:
```
with my_cond_dispatch_mode:
    make_fx()
```
just my 2 cents how to make this customizable, I might just be bragging.

Reply:
Sorry, I do not understand what you are suggesting. Why do we want to call `mode.__torch_dispatch__` explicitly here?

Reply:
Feel free to correct me, but from my understanding, if we register the `cond` op with `DispatchKey.Python`, shouldn't the contract be "trigger `__torch_dispatch__` when this op is called", instead of encoding a certain type of behavior inside the `python_fallback`?

Reply:
No sir, we handle the dispatch here - because we need to provide custom logic for it. Like, we could port most of this down to a proxy_call impl that checks name and then routes it back to here, but I am not sure that gives us any benefits.